### DISCUSSION TITLE
Use parametrisation to render many ipynb notebooks from a template

### DISCUSSION THREAD

**Discussion by danieltomasz** [quarto-dev/quarto-cli#9107] [Q&A] [answered]
### Description

I have a problem with getting parametrisation right.
My use case is: I do neuroimaging study, I would like to use jupyter notebook to run the same pipeline for data from all participant but on one step the analysis involve manual intervention to remove components that might differ from one participants to another

My idea is is have an `template.ipynb`(ipynb, not qmd)

with following cells:

```raw
title: "Test"
format: ipynb
jupyter: python3
execute:
  enabled: true
```
```python
# | tags: [parameters]

subject = 101
```
```python
print(subject)
```
and then create notebook for every participant where I could later adjust thing manually.
But I have a problem that the parametrisation doesnt work
running 
```
 quarto render notebooks/test/test.ipynb  -P subject:102
```
gives me a file `test.out.ipynb`  where `101` is not changed as a parameter 

I am on macOS 14.4 with quarto 1.5.23

**Comment by danieltomasz** [comment]
I am able to create a `ipynb` file from `qmd`, but not from `.ipynb` (after I install papermill to my jupyter env, which is not really documented here https://quarto.org/docs/computations/parameters.html)
````test.qmd
---
title: "Test"
format: ipynb
jupyter: conda-paths-3.12
execute:
  enabled: true
---

```{python}
# | tags: [parameters]

subject = 101

```

```{python}
print(subject)
```
````

and running 
```
quarto render notebooks/test/test.qmd -P subject:102 --to ipynb
```


In this setup in output file I have 2 cells where variable subject is defined 
![image](https://github.com/quarto-dev/quarto-cli/assets/7980381/78bcbfab-5618-4b3d-9eb4-e7b5aa8fcacd)

how can i make it only one cell? and how I can  ensure that parameter will occur in the output filename?

**Reply by danieltomasz** [reply, marked as answer]
`ipynb` works when I add the tag, but not via comment tag, but to cell via jupyter interface - in that regards the documentation was  confusing for me;

**Reply by mcanouil** [reply]
Would you mind elaborate on the confusing part of the documentation? What could be improved to avoid that for future users?
Thank for your feedback.

**Reply by danieltomasz** [reply]
For me most confusing was the part of tagging cell in  Jupyter 
according to the documentation https://quarto.org/docs/computations/parameters.html
>designate a cell with the tag parameters 

And in the cell below
```
#| tags: [parameters]
```
While what I actually needed to do was to install `papermill` and manually (through installing addition VSCode plugin `Jupyter Cell Tags`) add tag to the cell containing the parameter to substitute; just adding 
`#| tags: [parameters]` didnt have any effect when I was running `.ipynb` template

I was not  able to do make my template to work using what was explicitly in the documentation page (but mention of the `papermill` package set me on a good path)

**Reply by danieltomasz** [reply]
Also, the naming of the parametrized output — I would expect that quarto take control of it, but it created the same name of the output file for every parameter (I fixed it outside quarto by using an external make-like python package to run quarto for every parameter in a loop and substitute parameter in python fstring which I added as an output option); 
If I specified output as `ipynb` while the template was `ipynb`  (without adding the output option), quarto was overwriting the original file; I might create and issue with reproducible example, if this is a bug not a default behaviour

**Reply by cderv** [reply]
> While what I actually needed to do was to install papermill and manually (through installing addition VSCode plugin Jupyter Cell Tags) add tag to the cell containing the parameter to substitute; just adding
#| tags: [parameters] didnt have any effect when I was running .ipynb template

When executing a render with params, we do error when parameters are passed and papermill is not available
````
Starting python3 kernel...ERROR: 

The papermill package is required for processing --execute-params
````

In the doc, we only say (https://quarto.org/docs/computations/parameters.html#jupyter)
> For Jupyter, Quarto uses the same syntax for defining parameters as [Papermill](https://papermill.readthedocs.io/en/latest/usage-parameterize.html).

do you mean we should add a code cell for explicit installation step of papermill at this place in the doc ? 

> Also, the naming of the parametrized output — I would expect that quarto take control of it, but it created the same name of the output file for every parameter

This would be a feature request indeed. It is not something that quarto is doing right now.

**Reply by danieltomasz** [reply]
I would say that mentions/ stating more explicitly will suffice:
1) that the papermill need to be installed (or is needed) in the environment when running ipynb; in my opinion  the statement: "package X use syntax of package Y" doesn't imply " Package X needs the package Y in order to run specific processes and the syntax of Y could be used to define it" 
2) the tag 'parameters' need to be manually added to the cell; the current doc suggests that adding the textline into cell will suffice

**Reply by danieltomasz** [reply]
Without setting the right tag for the cell (which involved installing additional plugin in VSCode, having only the Jupyter plugin is not enough, this is different than using Jupyter Lab or Notebook, where the cell tags are available by default to the user) my template was treated as a "normal" notebook and the variable was not substituted; this took me most of my time to learn; I would say this need to be stated more explicit, as someone who didn't have experience with papermill I assumed that the usual "quarto" way of setting properties in the body of code cell is what the paragraph meant.

**Reply by mcanouil** [reply]
FYI, why did you not install the Jupyter extension pack? https://marketplace.visualstudio.com/items?itemName=ms-toolsai.jupyter
It comes with everything you need in order to work with Jupyter in VSCode.

> Jupyter plugin

Don't know what that is since in the marketplace `Jupyter` is the extension pack.

**Reply by danieltomasz** [reply]
@mcanouil I was using Jupyter  Extension before it was  shipped as extension pack  with the other extension (I remember separate installation of  extension pack to install but maybe this was enough long ago); maybe I uninstalled some extension, I don't remember the reason; VScode support for cells is less than 2 years ....

My issue is not how to install the Cell Tag extension, but that the text didn't suggest ( in my opinion) that I should use Jupyter cell tags and I  should manually add them;
>To parameterize a document, designate a cell with the tag parameters and provide appropriate default values:

If   I have cell extension installed and the  file `test.ipynb` have  the following content (same as in the docs):
```
#| tags: [parameters]
alpha = 0.1

```

and I will run 
```
quarto render test.ipynb -P alpha:0.2
```
The output will not have alpha parameter replaced and still have the value `alpha = 0.1` and dont produce any error.

For me is either bug (the tag in the cell is enough but is not picked by quarto to start  papermill process) or insufficient documentation about relying on external package (`papermill`) and the specific  setup it does need to be able to pick that the notebook is an template.

Thank for all devs for all the work! those comments are not complaining, but if I was confused (I use ipynb/rmd/qmd many years) more users can be confused as well (especially if they dont have experience with papermill package and its setup)

**Reply by danieltomasz** [reply]
And maybe I was missing the part  about adding
```
execute:
  enabled: true
```
from raw yaml at the top of ipynb, because without this running `quarto render` has no effect (even if I dont want to run the code, but only  to substitute the values in notebook)

**Reply by mcanouil** [reply]
> Thank for all devs for all the work! those comments are not complaining, but if I was confused (I use ipynb/rmd/qmd many years) more users can be confused as well (especially if they dont have experience with papermill package and its setup)

No arguing there. That's the main reason I asked you to elaborate in the first place.

**Reply by mcanouil** [reply]
> And maybe I was missing the part  about adding
> 
> ```
> 
> execute:
> 
>   enabled: true
> 
> ```
> 
> from raw yaml at the top of ipynb, because without this running `quarto render` has no effect (even if I dont want to run the code, but only  to substitute the values in notebook)

There is a flag to force cells to be executed, if I am not mistaken it is `--execute`.

**Reply by danieltomasz** [reply]
Indeed, but in my example example above without adding this flag (or via yaml) I wasnt able to substitute the parameters in the notebook; also this flag isnt mentioned in the doc page about parametrisation; the only flag mentioned is `--execute-params` which suggest loading params from yaml (which wasn't my case)
